### PR TITLE
feat: send event with fee rebates stats at end of epoch

### LINKED ISSUE #9261: [Feature]: Finalise referral set API (pt1)
# Feature Overview

**In order to** support fron end 
**We will** ensure the new referral set API covers all the below
**So that** front end can build
- The referrer stats on page 10 of the wireframes (and in the miro that will soon exist [here](https://miro.com/app/board/uXjVOs_PBKk=/?share_link_id=783456245937))

## Specs
https://github.com/vegaprotocol/specs/blob/cosmicelevator/protocol/0083-RFPR-on_chain_referral_program.md#apis

## Tasks
A checklist of the tasks that are needed to develop the feature and meet the acceptance criteria and feature test scenarios.
- [ ] Ensure the referral set API covers all of the things in the spec
  - [x] Team volume in last x days (referral_set_running_notional_taker_volume)
  - [x] Current tier (of the referrer)
  - [x] Total traders referred
  - [x] Total commission earned per asset per epoch
  - [x] Total discounts received by all referees per asset per epoch
  - [x] Members list
    - [x] Address
    - [x] Date joined (or number of epochs in team)

### Additional Details (optional)

Additional API updates to be completed in: https://github.com/vegaprotocol/vega/issues/9541

### PR DESCRIPTION
**Pull request by jeremyletang**
Addresses referral set statistics for #9261

close https://github.com/vegaprotocol/vega/issues/9261

### CODE REVIEW

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -6031,3 +6043,37 @@ type TeamRefereeHistoryConnection {
...
   pageInfo: PageInfo!
 }
+
+"Referral rewards and discounts that have been applied on a specific market/asset up to the given epoch."
+type ReferralSetFeeStats {
+  "The market the fees were paid in"
+  marketId: String!
+  "The settlement asset of the market."
+  assetId: String!
+  "The epoch for which these stats where valid."
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:6053]:
```suggestion
  "The epoch for which these stats were valid."
```

File: datanode/gateway/graphql/schema.graphql [hunk 2]
```diff
@@ -6031,3 +6043,37 @@ type TeamRefereeHistoryConnection {
...
+  referrerRewardsGenerated: [ReferrerRewardsGenerated!]!
+  "The total referral discounts applied to all referee taker fees"
+  refereesDiscountApplied: [PartyAmount!]!
+  "The total volume discounts applied to all referee taker fees"
+  volumeDiscountApplied: [PartyAmount!]!
+}
+
+"An amount received by a party as a reward or a discount"
+type PartyAmount {
+  "Id of the party paid"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:6067]:
```suggestion
  "ID of the party that paid"
```

File: datanode/gateway/graphql/schema.graphql [hunk 3]
```diff
@@ -6031,3 +6043,37 @@ type TeamRefereeHistoryConnection {
...
+  partyId: String!
+  "Amount received by the party"
+  amount: String!
+}
+
+"Rewards generated for referrers by each of their referees"
+type ReferrerRewardsGenerated {
+  "Id of the referral set's referrer"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:6075]:
```suggestion
  "ID of the referral set's referrer"
```

File: protos/sources/vega/events/v1/events.proto [hunk 4]
```diff
@@ -1035,6 +1069,9 @@ enum BusEventType {
   // Event indicating the updated statistics for the vesting program.
   BUS_EVENT_TYPE_VESTING_STATS_UPDATED = 77;
 
+  // Evnts indicating the fee stats per market at end of epoch
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1072]:
```suggestion
  // Events indicating the fee stats per market at end of epoch
```

File: protos/sources/vega/events/v1/events.proto [hunk 5]
```diff
@@ -1198,6 +1235,8 @@ message BusEvent {
     ReferralSetStatsUpdated referral_set_stats_updated = 174;
     // Event notifying of an update to the vesting statistics.
     VestingStatsUpdated vesting_stats_updated = 175;
+    // Event notifying of an update the fees stats for a market.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1238]:
```suggestion
    // Event notifying of an update to the fees stats for a market.
```

File: protos/sources/vega/events/v1/events.proto [hunk 6]
```diff
@@ -29,6 +29,40 @@ message PartyVestingStats {
...
+
+// Rewards generated for referrers by each of their referees
+message ReferrerRewardsGenerated {
+  // The referrer party
+  string referrer = 1;
+  // The amount of rewards generated per party.
+  repeated PartyAmount generated_reward = 2;
+}
+
+// A pair of a party and amount
```
[the remark this answers is not in the document]
Comment (guoguojin) [reply protos/sources/vega/events/v1/events.proto:58]: Yes, it's only used in the fee stats, and the referrer rewards generated.

**Review by candida-d — DISMISSED** [review]

File: core/execution/spot/market_snapshot.go [hunk 7]
```diff
@@ -199,6 +208,7 @@ func (m *Market) GetState() *types.ExecSpotMarket {
 		Parties:                    parties,
 		Closed:                     m.closed,
 		HasTraded:                  m.hasTraded,
+		FeeStats:                   m.fee.GetState(),
 	}
 
 	println(em.IntoProto().String())
```
Comment (EVODelavega) [inline core/execution/spot/market_snapshot.go:214]: debug statement? `println` should not be used

**Review by EVODelavega — DISMISSED** [review]

File: core/execution/spot/market_snapshot.go (hunk 7 above)
Comment (guoguojin) [reply core/execution/spot/market_snapshot.go]: Thanks, removed now.

**Review by EVODelavega — DISMISSED** [review]

File: datanode/gateway/graphql/schema.graphql [hunk 8]
```diff
@@ -812,6 +812,16 @@ type Query {
     pagination: Pagination
   ): ReferralSetConnection!
 
+  "Get referrer fee and discount stats"
+  referralFeeStats(
+    "Optional market ID to filter for. If omitted, asset ID must be provided."
+    marketID: ID
```
Comment (mattrussell36) [inline datanode/gateway/graphql/schema.graphql:818]: Where not perfect everywhere on this, so consider it non blocking but `marketId` lowercase 'd' is more consistent with the rest of the schema

**Review by mattrussell36 — DISMISSED** [review]

**Review by asiaznik — DISMISSED** [review]

File: datanode/gateway/graphql/schema.graphql (hunk 8 above)
Comment (guoguojin) [reply datanode/gateway/graphql/schema.graphql]: I had to fix a number of merge conflicts so I have updated this along with the assetId which also suffered from the same capitalisation.

**Pull request merged**